File: src/player.js
import * as THREE from 'three';
import { groundHeight, slopeAt, obstacleAt } from './terrain.js';

export const PLAYER_HALF_HEIGHT = 0.5;
const GRAVITY = -26;
const MIN_SPEED = 9;
const MAX_SPEED = 26;
const SLOPE_ACCEL = 34; // how strongly downhill slope speeds you up
const JUMP_BASE = 8;
const JUMP_CHARGE_MAX = 6.5; // added on top of JUMP_BASE at full charge
const MAX_CHARGE_TIME = 0.55; // seconds of holding to reach full charge
const RAMP_LAUNCH_BOOST = 11;
const SPIN_SPEED = Math.PI * 2.2; // radians/sec while spinning
const LANDING_TOLERANCE = 0.35; // radians of allowed rotation error on landing

const BODY_COLOR = 0x2b3140;
const BOARD_COLOR = 0xff8a3d;

function buildRiderMesh() {
  const group = new THREE.Group();

  const boardGeo = new THREE.PlaneGeometry(1.5, 0.22);
  const boardMat = new THREE.MeshBasicMaterial({ color: BOARD_COLOR });
  const board = new THREE.Mesh(boardGeo, boardMat);
  board.position.set(0, -0.4, 0.05);
  group.add(board);

  const bodyShape = new THREE.Shape();
  bodyShape.moveTo(-0.28, -0.3);
  bodyShape.lineTo(-0.22, 0.35);
  bodyShape.quadraticCurveTo(0, 0.55, 0.22, 0.35);
  bodyShape.lineTo(0.28, -0.3);
  bodyShape.closePath();
  const bodyGeo = new THREE.ShapeGeometry(bodyShape);
  const bodyMat = new THREE.MeshBasicMaterial({ color: BODY_COLOR });
  const body = new THREE.Mesh(bodyGeo, bodyMat);
  body.position.set(0, 0, 0.06);
  group.add(body);

  const headGeo = new THREE.CircleGeometry(0.22, 16);
  const head = new THREE.Mesh(headGeo, bodyMat);
  head.position.set(0, 0.65, 0.06);
  group.add(head);

  return group;
}

export class Player {
  constructor() {
    this.x = 0;
    this.y = groundHeight(0) + PLAYER_HALF_HEIGHT;
    this.vy = 0;
    this.speed = MIN_SPEED + 3;
    this.grounded = true;
    this.rotation = 0;
    this.spinVelocity = 0;
    this.totalRotation = 0;
    this.state = 'running'; // running | airborne | crashed | finished
    this.chargeTime = 0;
    this.charging = false;
    this.mesh = buildRiderMesh();
    this._lastRampX = null;
  }

  startCharge() {
    if (this.state !== 'running' && this.state !== 'airborne') return;
    if (this.grounded) {
      this.charging = true;
      this.chargeTime = 0;
    } else {
      // Airborne tap: begin/continue a trick spin.
      this.spinVelocity = SPIN_SPEED;
    }
  }

  releaseCharge() {
    if (this.charging && this.grounded) {
      const t = Math.min(this.chargeTime, MAX_CHARGE_TIME);
      const power = t / MAX_CHARGE_TIME;
      this.vy = JUMP_BASE + JUMP_CHARGE_MAX * power;
      this.grounded = false;
      this.state = 'airborne';
      this.totalRotation = 0;
    }
    this.charging = false;
    this.chargeTime = 0;
  }

  /** Advances physics by dt seconds. Returns an event object describing what happened. */
  update(dt) {
    const events = { crashed: false, landed: false, trickBonus: 0 };
    if (this.state === 'crashed' || this.state === 'finished') return events;

    if (this.charging && this.grounded) {
      this.chargeTime += dt;
    }

    const slope = slopeAt(this.x);
    this.speed += -slope * SLOPE_ACCEL * dt;
    this.speed = Math.max(MIN_SPEED, Math.min(MAX_SPEED, this.speed));
    this.x += this.speed * dt;

    const ground = groundHeight(this.x);

    if (this.grounded) {
      this.y = ground + PLAYER_HALF_HEIGHT;
      this.rotation = Math.atan(slopeAt(this.x));

      const obstacle = obstacleAt(this.x);
      if (obstacle && obstacle.type === 'rock') {
        this.state = 'crashed';
        events.crashed = true;
        return events;
      }
      if (obstacle && obstacle.type === 'ramp' && this._lastRampX !== obstacle.x) {
        this._lastRampX = obstacle.x;
        this.vy = RAMP_LAUNCH_BOOST;
        this.grounded = false;
        this.state = 'airborne';
        this.totalRotation = 0;
        this.charging = false;
      }
    } else {
      this.vy += GRAVITY * dt;
      this.y += this.vy * dt;

      if (this.spinVelocity !== 0) {
        const delta = this.spinVelocity * dt;
        this.rotation += delta;
        this.totalRotation += delta;
      }

      const floor = ground + PLAYER_HALF_HEIGHT;
      if (this.y <= floor && this.vy <= 0) {
        this.y = floor;
        this.vy = 0;
        this.grounded = true;
        this.spinVelocity = 0;

        const obstacle = obstacleAt(this.x);
        if (obstacle && obstacle.type === 'rock') {
          this.state = 'crashed';
          events.crashed = true;
          return events;
        }

        const spins = this.totalRotation / (Math.PI * 2);
        const nearestWhole = Math.round(spins);
        const error = Math.abs(spins - nearestWhole) * Math.PI * 2;
        const attemptedTrick = Math.abs(this.totalRotation) > 0.5;

        if (attemptedTrick && error <= LANDING_TOLERANCE && nearestWhole !== 0) {
          events.trickBonus = Math.abs(nearestWhole) * 50;
          this.rotation = 0;
        } else if (attemptedTrick) {
          // Wipeout: bad landing costs speed but the run continues.
          this.speed = Math.max(MIN_SPEED, this.speed * 0.35);
          this.rotation = 0;
        } else {
          this.rotation = Math.atan(slopeAt(this.x));
        }

        this.state = 'running';
        events.landed = true;
        this.totalRotation = 0;
      }
    }

    this.mesh.position.set(this.x, this.y, 0.5);
    this.mesh.rotation.z = this.rotation;

    return events;
  }
}

File: public/py/player.py
"""Snowboarder physics for Sky's Adventure - pure Python, no rendering.

Runs inside Pyodide (WebAssembly) in the browser. JavaScript reads the
resulting x/y/rotation/state each frame (via game.py) to update the
Three.js mesh; all movement, gravity, jumping and trick math happens here.
"""
import math
import terrain

PLAYER_HALF_HEIGHT = 0.5
GRAVITY = -26
MIN_SPEED = 9
MAX_SPEED = 26
SLOPE_ACCEL = 34  # how strongly downhill slope speeds you up
JUMP_BASE = 8
JUMP_CHARGE_MAX = 6.5  # added on top of JUMP_BASE at full charge
MAX_CHARGE_TIME = 0.55  # seconds of holding to reach full charge
RAMP_LAUNCH_BOOST = 11
SPIN_SPEED = math.pi * 2.2  # radians/sec while spinning
LANDING_TOLERANCE = 0.35  # radians of allowed rotation error on landing


class Player:
    def __init__(self):
        self.x = 0.0
        self.y = terrain.ground_height(0) + PLAYER_HALF_HEIGHT
        self.vy = 0.0
        self.speed = MIN_SPEED + 3
        self.grounded = True
        self.rotation = 0.0
        self.spin_velocity = 0.0
        self.total_rotation = 0.0
        self.state = "running"  # running | airborne | crashed | finished
        self.charge_time = 0.0
        self.charging = False
        self._last_ramp_x = None

    def start_charge(self):
        if self.state not in ("running", "airborne"):
            return
        if self.grounded:
            self.charging = True
            self.charge_time = 0.0
        else:
            # Airborne tap: begin/continue a trick spin.
            self.spin_velocity = SPIN_SPEED

    def release_charge(self):
        if self.charging and self.grounded:
            t = min(self.charge_time, MAX_CHARGE_TIME)
            power = t / MAX_CHARGE_TIME
            self.vy = JUMP_BASE + JUMP_CHARGE_MAX * power
            self.grounded = False
            self.state = "airborne"
            self.total_rotation = 0.0
        self.charging = False
        self.charge_time = 0.0

    def update(self, dt):
        """Advances physics by dt seconds. Returns an events dict."""
        events = {"crashed": False, "landed": False, "trick_bonus": 0}
        if self.state in ("crashed", "finished"):
            return events

        if self.charging and self.grounded:
            self.charge_time += dt

        slope = terrain.slope_at(self.x)
        self.speed += -slope * SLOPE_ACCEL * dt
        self.speed = max(MIN_SPEED, min(MAX_SPEED, self.speed))
        self.x += self.speed * dt

        ground = terrain.ground_height(self.x)

        if self.grounded:
            self.y = ground + PLAYER_HALF_HEIGHT
            self.rotation = math.atan(terrain.slope_at(self.x))

            obstacle = terrain.obstacle_at(self.x)
            if obstacle and obstacle["type"] == "rock":
                self.state = "crashed"
                events["crashed"] = True
                return events
            if obstacle and obstacle["type"] == "ramp" and self._last_ramp_x != obstacle["x"]:
                self._last_ramp_x = obstacle["x"]
                self.vy = RAMP_LAUNCH_BOOST
                self.grounded = False
                self.state = "airborne"
                self.total_rotation = 0.0
                self.charging = False
        else:
            self.vy += GRAVITY * dt
            self.y += self.vy * dt

            if self.spin_velocity != 0:
                delta = self.spin_velocity * dt
                self.rotation += delta
                self.total_rotation += delta

            floor = ground + PLAYER_HALF_HEIGHT
            if self.y <= floor and self.vy <= 0:
                self.y = floor
                self.vy = 0.0
                self.grounded = True
                self.spin_velocity = 0.0

                obstacle = terrain.obstacle_at(self.x)
                if obstacle and obstacle["type"] == "rock":
                    self.state = "crashed"
                    events["crashed"] = True
                    return events

                spins = self.total_rotation / (2 * math.pi)
                nearest_whole = round(spins)
                error = abs(spins - nearest_whole) * 2 * math.pi
                attempted_trick = abs(self.total_rotation) > 0.5

                if attempted_trick and error <= LANDING_TOLERANCE and nearest_whole != 0:
                    events["trick_bonus"] = abs(nearest_whole) * 50
                    self.rotation = 0.0
                elif attempted_trick:
                    # Wipeout: bad landing costs speed but the run continues.
                    self.speed = max(MIN_SPEED, self.speed * 0.35)
                    self.rotation = 0.0
                else:
                    self.rotation = math.atan(terrain.slope_at(self.x))

                self.state = "running"
                events["landed"] = True
                self.total_rotation = 0.0

        return events
